[Github Actions] Add new workflows for fullnode syncing.

JoshLind opened the pull request:
### Description
This PR adds nightly Github action workflows to verify fullnode health. Specifically, we add the following workflows:
1. Start a node using the `main` branch, connect it to `devnet` and verify it can sync using execution syncing.
2. Start a node using the `main` branch, connect it to `testnet` and verify it can sync using fast syncing.
3. Start a node using the `main` branch, connect it to `mainnet` and verify it can sync using output syncing.
4. Start a node using the `main` branch, connect it to `mainnet` and verify it can sync using fast syncing.
5. Start a node using the `main` branch, connect it to `devnet` and verify it can sync using intelligent syncing mode.


The workflows will spawn and monitor the node and wait for it to sync to a target version (where the target is the latest version of the blockchain plus some delta). If this fails, the workflow will fail.

We'll likely add more of these tests in the future, but let's see how these do.

### Test Plan
Manual verification.

rustielin commented on .github/actions/fullnode-sync/action.yaml:
```diff
@@ -0,0 +1,85 @@
+# This action runs a public fullnode using a specified branch ($BRANCH),
```
nit: I would make the environment variables `BRANCH`, `NETWORK`, `BOOTSTRAPPING_MODE`, etc into action inputs, which makes it slightly more modular and clearer to the user what's required. Also, it makes it easier later on to move all these actions to a separate repo! Here's an example (from me moving out the replay-verify job) : https://github.com/aptos-labs/actions/blob/0a510a8f270c9b5093c3e79043f4613dbb07aeb3/replay-verify/action.yml

rustielin reviewed (changes requested):
Awesome work! Generally LGTM, but some nits on the structure of the composite action, as well as to split out the main action to a separate script

rustielin commented on .github/actions/fullnode-sync/action.yaml:
```diff
@@ -0,0 +1,85 @@
+        cp config/src/config/test_data/public_full_node.yaml .
+        
+        # Update the working directory (from /opt/aptos/data to /tmp/)
+        sed -i -e 's/\/opt\/aptos\/data/\/tmp\//g' public_full_node.yaml
```
I'm usually the worst offender of this :) but can we split most of the meat of the logic here into a separate script? That way it's easier to test and verify locally. Also gets us away from bash mess/magic

rustielin commented on .github/workflows/fullnode-execute-devnet-main.yaml:
```diff
@@ -0,0 +1,51 @@
+      GIT_SHA:
+        required: false
+        type: string
+        description: The git SHA1 to test. If not specified, Forge will check the latest commits on the current branch
```
this input is not actually used -- can you remove it?

JoshLind replied:
Copy/pasta magic that I thought might be used. Removed from them all. 😄

JoshLind commented:
Thanks @rustielin! Made a few tweaks:
- Introduced a new python script, which is a bit more readable and easier to manager.
- Added some features, e.g., verifying the process is still running, handling timeouts if the node appears to be stuck, measuring syncing throughput, etc.
- Added composite inputs and a README, like you did: https://github.com/aptos-labs/actions/pull/3/files
- Hopefully, once your PR lands, it won't be too hard to migrate this across.

rustielin commented on .github/actions/fullnode-sync/fullnode_sync.py:
```diff
@@ -0,0 +1,267 @@
+  # Clone the aptos-networks repository at the same place as aptos-core
+  current_working_directory = os.getcwd()
+  os.chdir("..")
+  subprocess.run(["git", "clone", "https://github.com/aptos-labs/aptos-networks"])
```
nit: instead of git clone, you can prob just get it from the raw paths here: e.g. https://raw.githubusercontent.com/aptos-labs/aptos-networks/main/mainnet/waypoint.txt
That way, you can get around having to do git magic and chdir ..

JoshLind replied:
Nice! So much simpler 😄

rustielin commented on .github/actions/fullnode-sync/fullnode_sync.py:
```diff
@@ -0,0 +1,267 @@
+def ping_rest_api_index_page(rest_endpoint, exit_if_none):
+  """Pings and returns the index page from a REST API endpoint"""
+  # Ping the REST API index page
+  process = subprocess.Popen(["curl", rest_endpoint], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```
instead of `curl`, probably use https://pypi.org/project/requests/, or at least do `curl -s`

Looks like job output gets noisy with
```
b''
Found output on stderr for ping_rest_api_index_page: b'  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n                                 Dload  Upload   Total   Spent    Left  Speed\n\r  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\r100   221  100   221    0     0   215k      0 --:--:-- --:--:-- --:--:--  215k\n'
```

JoshLind replied:
Nice, thanks! Just gone for `curl -s` for now 😄

bchocho reviewed (approved):
This is incredible. I saw the max time is 5 hours. How long does the sync take today? In future could this be the basis of a regression test?
